include/pybind11/native_enum.h

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2022 The pybind Community.
+// All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#pragma once
+
+#include "pybind11.h"
+
+#include <type_traits>
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+/// Conversions between Python's native (stdlib) enum types and C++ enums.
+template <typename Type>
+class native_enum {
+public:
+    using Underlying = typename std::underlying_type<Type>::type;
+    // Scalar is the integer representation of underlying type
+    using Scalar = detail::conditional_t<detail::any_of<detail::is_std_char_type<Underlying>,
+                                                        std::is_same<Underlying, bool>>::value,
+                                         detail::equivalent_integer_t<Underlying>,
+                                         Underlying>;
+
+    template <typename... Extra>
+    native_enum(const handle &scope, const char *name, const Extra &.../*extra*/)
+        : m_scope(reinterpret_borrow<object>(scope)), m_name(name) {
```

Comment:
If you are storing the scope anyway, why not allow it take an object in a delegating ctor?

Reply:
I'm only at the stage of trying to get the basic mechanisms in place. I'm thinking in the final version I may need to make `native_enum` a function, to avoid the dtor trick (will kill the process if something goes wrong).

At this point I'm emulating the existing `enum_` interface as much as possible.